### DISCUSSION TITLE
homework3: Unreachable path of FSM and missing clip for jump_end_from_walk_run

### DISCUSSION THREAD

**Discussion by huandzh** [BoomingTech/Piccolo#259] [General]
_The bug? report is about homework03's task description but may require some revision in its branch._ 

**Describe the bug | Bug描述**

The task description of homework03 may contain the following issues:

1. There is one unreachable path of the FSM diagram, since `start_walk_end` is always false in the code skeleton.  
1. `jump_end_from_idle` and `jump_end_from_walk_run` use the same clip(`jump_stop`). `jump_end_from_walk_run` should have a clip which ends earlier. Else, the character will float after running and jumping. 

Students may be unclear about whether they are required to fix them.

**Expected behavior | 预期行为**

* For FSM issue: change diagram and code skeleton or require students to fix it
* For clip issue: add clips accordingly?

Or state the current issues in task description.
    
**Screenshots/Logs | 截图/日志**
![image](https://user-images.githubusercontent.com/1454871/172667318-00357038-bb22-4d36-a3d3-fcf04b8b8712.png)

![image](https://user-images.githubusercontent.com/1454871/172668424-c11fb3eb-d33e-4d0d-972a-97b546e8fb94.png)

**Comment by yijiang233** [comment]
Thank you for your feedback!
For issue 1:
start_walk_end needs to be determined by several factors, i.e. current speed, playing state of animation clip, and which clip is playing. Even considering these factors, current animation resource still feels creepy and confusing. So to lower the homework difficulty, start_walk_end was set to false.
Without affecting the implementation of ASM, walk_run can only switch to jump_start_from_walk_run with priority 1 and idle with priority 3.
Students with interests can implement the logic for start_walk_run for the robot stop playing animation at suitable timing.
由于start_walk_end需要根据当前速度、动画播放位置、当前播放的动画片段等因素决定是否播放停止动画。即使考虑这些因素使用当前素材仍然会使动画不自然，使同学们困惑。因此为了降低难度在前面直接把start_walk_end置为false。
在不影响状态机的实现逻辑的前提下使walk_run只能进入优先级为1的jump_start_from_walk_run或者优先级为3的idle。
有能力的同学可以修改start_walk_run的计算逻辑使小白人在合适时机播放停止动画。
For issue 2: 
You're right about the correct design of the animation clips. More animation clips can help improving the performance of the animation. Currently there is no plan on making more animation clips.
你说的没错，加入更多的动画资源能够有更好的表现效果，暂时没有计划做更多的资源。

**Comment by huandzh** [comment]
Thanks for your reply. 

These comments are helpful for other students. We may keep the issue open for a while.

**Comment by Ol6rin** [comment]
I will transfer this issue to discussion

**Reply by huandzh** [reply]
Thanks!
